### PR TITLE
Improve thread safety of LeaderElection algorithm (redux)

### LINKED ISSUE #59: LeaderElectionBase is not thread safe
**Describe the bug**

`LeaderElectionBase` is not thread safe.

- If two threads call `StartAsync()` at the same time on a stopped LeaderElection instance, then it is possible for both threads to create a new `_leadershipLoopCancellationTokenSource` and `_leadershipLoopTask`. The first assigned task will be orphaned by the second assignment, and that orphaned task will execute in the background for the lifetime of the application. 

- `StopAsync()` suffers from a **Check-then-Act** vulnerability, specifically inside the "if field not null" expressions. Two threads/tasks could enter these blocks at the same time, then one might pause while the other completes the function and set the member to null, resulting in a null reference exception when the other thread/task resumes.

- The public `TryAcquireLeadershipAsync()` method contains a potential deadlock. This method invokes the `LeadershipChanged` event while the `_leadershipSemaphore` is acquired. If the user's callback was to somehow call `TryAcquireLeadershipAsync()` again, then a deadlock will occur.

**Expected behavior**

- [ ] The state of a LeaderElection instance should be thread safe.
  - In the example given above, the first thread should block the second thread until it completes all work on the internal state.
  - Read-only properties like `IsLeader` and `LastLeadershipRenewal` do not require the critical section as long as the code is written such that these fields can be safely accessed outside of the critical section, by strategic use of Volatile and Interlocked operations.
- [ ] Callback events should be invoked **outside** of the critical section to prevent deadlocks and other potential reentrancy issues. They should be executed **after** all internal state has been updated.
- [ ] Derived implementations should not _need_ to worry about thread-safety. In other words, the base implementation should guarantee that the protected virtual methods are called within a thread-safe context. This implies that derived implementations **must not** fire Events.

### LINKED ISSUE #60: LeaderElectionBase.StartAsync() links the leader-loop task to the supplied cancellation token
**Describe the bug**

The `LeaderElectionBase.StartAsync(CancellationToken)` method links the `_leadershipLoopCancellationTokenSource` to the provided cancellation token. This allows the caller to cancel the leader-loop task without the LeaderElection instances' knowledge. This is wrong. Unless explicitly documented otherwise, the provided cancellation token should only be used to cancel the `StartAsync()` method itself. 

The user must call `StopAsync()` or `DisposeAsync()` to cancel the leader-loop task.

There is also no reason to expect that the provided cancellation token is long-lived (it isn't documented anywhere). In many cases tokens are short-lived, such as those created when processing an ASP.NET API request, which is disposed as soon as the request completes.

**Expected behavior**

- [ ] The provided cancellation token is only used to cancel the `StartAsync()` method itself.
- [ ] The provided cancellation token has no effect on the lifetime of the leader-loop task.

### PR DESCRIPTION
**Pull request by mrfootoyou**
## Description

> [!IMPORTANT]
> This PR is a copy of #65 which was merged into wrong base branch.

> [!IMPORTANT]
> This PR contains functional changes. See the review comments for details. 

This pull request refactors `BaseLeaderElection` to ensure robust concurrent access to leadership status and the background loop by replacing volatile fields with explicit Volatile and Interlocked operations. This change prevents potential race conditions and torn reads, particularly for leadership renewal timestamps, while centralizing state transitions to maintain consistency across the lifecycle of the leader election instance. This resolves #59

It resolves #60 by **not** tying the `_leaderLoopTaskCTS` to the `StartAsync()` cancellation token.

It also introduces several improvements and clarifications to the leader election interface, its documentation, and improves test coverage. It includes enhanced XML documentation for the `ILeaderElection` interface, the addition of new properties and exception details, and a significant expansion of unit tests to cover edge cases and error handling. Additionally, some package dependencies have been updated.

---

## Checklist

- [x] My code follows the project's coding style
- [x] I have self-reviewed my changes
- [x] I have added tests (or explain why not needed)
- [ ] I have updated the documentation
- [x] I have linked relevant issue(s)
- [x] I have included meaningful commit messages

---

## Related Issues

Closes #59 and #60

---

## Additional Notes

Note that the target branch is `refactor/di-extensions` and should not be merged until after #64 is merged.

### CODE REVIEW

File: src/LeaderElection/LeaderElectionBase.cs [hunk 1]
```diff
@@ -19,252 +46,580 @@ public abstract partial class LeaderElectionBase<TSettings> : ILeaderElection
...
         {
-            LogStoppingLeaderElectionForInstanceInstanceid(_settings.InstanceId);
-
-            await _leadershipLoopCancellationTokenSource.CancelAsync().ConfigureAwait(false);
+            return; // already stopped
         }
 
-        if (_leadershipLoopTask != null)
+        var isLeader = await AcquireLeaderLoopSemaphoreAsync(cancellationToken)
+            .ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (mrfootoyou) [reply src/LeaderElection/LeaderElectionBase.cs:495]: Done

File: src/LeaderElection/LeaderElectionSettingsBase.cs [hunk 2]
```diff
@@ -15,17 +15,57 @@ public class LeaderElectionSettingsBase
     public string InstanceId { get; set; } = Guid.NewGuid().ToString();
 
     /// <summary>
-    /// The interval at which the leader will attempt to renew its leadership.
+    /// The interval at which non-leader instances will attempt to acquire leadership.
     /// <para/>
     /// Must be greater than or equal to zero.
```
[the remark this answers is not in the document]
Comment (mrfootoyou) [reply src/LeaderElection/LeaderElectionSettingsBase.cs]: Done

**Comment by mrfootoyou** [comment]
@gregyjames I merged the other PR into the di-extensions branch. This duplicate PR fixes.

**Review by gregyjames — APPROVED** [review]
